### PR TITLE
feat: fw-904 add bookmark

### PR DESCRIPTION
**Pull request by swkatmask**
https://mask.atlassian.net/browse/fw-904

### CODE REVIEW

File: src/app/(normal)/bookmark/BookmarkList.tsx [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    source: SocialSource;
+}
+
+export function BookmarkList({ source }: Props) {
+    console.log('sources', source);
+    const query = useSuspenseInfiniteQuery({
+        queryKey: ['posts', source, 'bookmark'],
+        queryFn: async ({ pageParam }) => {
+            const provider = resolveSocialMediaProvider(source);
+            if (!provider) return;
```
Comment (guanbinrui) [inline src/app/(normal)/bookmark/BookmarkList.tsx:22]:
```suggestion
```

File: src/hooks/useToggleBookmark.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+import { enqueueErrorMessage, enqueueSuccessMessage } from '@/helpers/enqueueMessage.js';
+import { resolveSocialMediaProvider } from '@/helpers/resolveSocialMediaProvider.js';
+import type { Post } from '@/providers/types/SocialMedia.js';
+
+export function useToggleBookmark() {
+    return useMutation({
+        mutationFn: async (post: Post) => {
+            const provider = resolveSocialMediaProvider(post.source);
+            if (!provider) return;
```
Comment (guanbinrui) [inline src/hooks/useToggleBookmark.ts:12]:
```suggestion
```

File: src/providers/firefly/SocialMedia.ts [hunk 3]
```diff
@@ -671,6 +672,19 @@ class FireflySocialMedia implements Provider {
     async getPostsQuoteOn(postId: string, indicator?: PageIndicator): Promise<Pageable<Post, PageIndicator>> {
         throw new Error('Method not implemented.');
     }
+
+    async getBookmarks(indicator?: PageIndicator): Promise<Pageable<Post, PageIndicator>> {
+        const url = urlcat(FIREFLY_ROOT_URL, '/v1/bookmark/find', {
+            post_type: 'all',
```
Comment (guanbinrui) [inline src/providers/firefly/SocialMedia.ts:678]:
```suggestion
            post_type: BookmarkType.All,
```

File: src/providers/firefly/SocialMedia.ts [hunk 4]
```diff
@@ -671,6 +672,19 @@ class FireflySocialMedia implements Provider {
...
+            platforms: 'farcaster',
+            limit: 25,
+            cursor: indicator?.id || undefined,
+        });
+        const response = await farcasterSessionHolder.fetch<BookmarkResponse>(url);
+        console.log('TODO firefly bookmark list', response);
```
Comment (guanbinrui) [inline src/providers/firefly/SocialMedia.ts:684]:
```suggestion
```

File: src/providers/firefly/SocialMedia.ts [hunk 5]
```diff
@@ -671,6 +672,19 @@ class FireflySocialMedia implements Provider {
     async getPostsQuoteOn(postId: string, indicator?: PageIndicator): Promise<Pageable<Post, PageIndicator>> {
         throw new Error('Method not implemented.');
     }
+
+    async getBookmarks(indicator?: PageIndicator): Promise<Pageable<Post, PageIndicator>> {
+        const url = urlcat(FIREFLY_ROOT_URL, '/v1/bookmark/find', {
+            post_type: 'all',
+            platforms: 'farcaster',
```
Comment (guanbinrui) [inline src/providers/firefly/SocialMedia.ts:679]: Use a proper enum.

Comment (swkatmask) [reply src/providers/firefly/SocialMedia.ts:679]: I don't think it's appropriate to use `SourceInURL`, they just happen to be the same value, but they are not the same concept.

Changes in `SourceInURL` could break this unexpectedly.

**Pull request merged**